Add zig tree-sitter support

voroskoi opened the pull request:
Highlighting works, indenting not so much, but I guess it is more than nothing.
Queries are from https://github.com/Himujjal/tree-sitter-zig

kirawi commented:
`indents.scm` and `folds.scm` are not supported. For indents you'll have to define a `indents.toml` instead, which looks like this: https://github.com/helix-editor/helix/blob/master/runtime/queries/rust/indents.toml

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,191 @@
+;; other identifiers
+(type_identifier) @type
+(primitive_type) @type.builtin
+(field_identifier) @field
```
`field` is not a scope we use, it should be `property`.

voroskoi replied:
How do I know which scopes are available? I have no idea where these come from.

kirawi replied:
Ah, it's listed here: https://docs.helix-editor.com/themes.html

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,191 @@
+
+(call_modifier) @keyword ; async
+
+(binary_operator) @keyword.operator
```
Change this to just `operator`.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,191 @@
+(integer_literal) @number
+(float_literal) @number
+
+(boolean_literal) @boolean
```
This should be `constant.builtin`.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,191 @@
+  "else"
+  "if"
+  "switch"
+] @conditional
```
This should be `keyword.control`. I'm not sure whether it would make sense to change it to `keyword.control.conditional`, I'm concerned about having unstandardized scopes @archseer .

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,191 @@
+[
+  "for"
+  "while"
+] @repeat
```
This should also be `keyword.control`.

voroskoi commented:
Thank You for the heads up, I have cleaned up things a bit. I have switched to a different grammar repo, but nvim-treesitter also use this one, so it is probably more up to date. Still have to figure out the indenting, does not work for me.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+)
+
+(ParamDecl 
+  (ParamType (SuffixExpr (IDENTIFIER) @parameter))
```
should be `@variable.parameter`

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+)
+
+(ParamDecl 
+  (IDENTIFIER) @parameter
```
Should also be `@variable.parameter`

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+(BUILTINIDENTIFIER) @function.builtin
+
+((BUILTINIDENTIFIER) @include
+  (#any-of? @include "@import" "@cImport"))
```
These should probably just be `@keyword`. I wonder if `#any-of?` is supported.

voroskoi replied:
Removed this one, does not matter.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  (STRINGLITERALSINGLE)
+] @string
+
+(CHAR_LITERAL) @character
```
Should be `@string`

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  "else"
+  "if"
+  "switch"
+] @keyword.control
```
These should be collapsed into one block.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  "return"
+  "break"
+  "continue"
+] @keyword
```
These should be `@keyword.control`

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  "resume"
+  "export"
+  "extern"
+] @function.macro
```
I'm not familiar with Zig, but are these related macros? They sound like a mix of `@keyword` and `@keyword.control`

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+[
+  ".."
+  "..."
+] @punctuation
```
These should also be collapsed into one block.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  "pub"
+  "usingnamespace"
+] @keyword
+
```
I would suggest collapsing all `@keyword` matches into one block as well.

kirawi commented on runtime/queries/zig/injections.scm:
```diff
@@ -0,0 +1,5 @@
+  (container_doc_comment)
+  (doc_comment)
+  (line_comment)
+] @comment
```
I'm not sure that this would work, since you also have to specify the injection language like here: https://github.com/helix-editor/helix/blob/master/runtime/queries/rust/injections.scm

voroskoi replied:
I just remove them for now, does not really matter.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  ".?"
+  ".*"
+  "="
+] @operator
```
These should also be grouped up with the other `@operator` blocks into one `@operator` block.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,277 @@
+  (line_comment)
+] @comment
+
+(IDENTIFIER) @variable
```
Does highlighting really work? I was under the assumption that it would highlighting everything as a variable at this point (since I don't think you can override them). I might be wrong since `locals.scm` is involved.

voroskoi replied:
No, it does not, You are right.

kirawi reviewed (changes requested):
You're getting there!

voroskoi commented:
Ok, I think the highlights part is fine for initial support. Time to figure out the indenting part.

kirawi commented on runtime/queries/zig/highlights.scm:
```diff
@@ -0,0 +1,250 @@
+  )
+)
+
+(BUILTINIDENTIFIER) @function.builtin
```
This should also be in the `@function.builtin` block.

voroskoi replied:
Done.

voroskoi commented:
I have removed the `indents.toml` as I can not figure out why it does not do anything.

kirawi commented:
It seems like the highlighting is a little incorrect:
![Capture](https://user-images.githubusercontent.com/67773714/130362246-73b0ae1a-91ab-4e9c-b120-bb067af375f8.PNG)

Function calls are being highlighted as variables.

archseer commented:
> I have removed the `indents.toml` as I can not figure out why it does not do anything.

Can you show me the file (it's gone from git history)? I might be able to help there.

archseer commented:
Nice work! I'll do some testing on indents but I think it's good for merge 👍
